File: services/event-streaming-service/src/models/topology.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::{FromRow, types::Json as SqlJson};
use uuid::Uuid;

use super::{
    sink::{
        BackpressureSnapshot, CepMatch, ConnectorCatalogEntry, LiveTailEvent, StateStoreSnapshot,
        WindowAggregate,
    },
    stream::ConnectorBinding,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopologyNode {
    pub id: String,
    pub label: String,
    pub node_type: String,
    pub stream_id: Option<Uuid>,
    pub window_id: Option<Uuid>,
    pub config: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopologyEdge {
    pub source_node_id: String,
    pub target_node_id: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JoinDefinition {
    pub join_type: String,
    pub left_stream_id: Uuid,
    pub right_stream_id: Uuid,
    pub table_name: String,
    pub key_fields: Vec<String>,
    pub window_seconds: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CepDefinition {
    pub pattern_name: String,
    pub sequence: Vec<String>,
    pub within_seconds: i32,
    pub output_stream: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackpressurePolicy {
    pub max_in_flight: i32,
    pub queue_capacity: i32,
    pub throttle_strategy: String,
}

impl Default for BackpressurePolicy {
    fn default() -> Self {
        Self {
            max_in_flight: 512,
            queue_capacity: 2048,
            throttle_strategy: "credit-based".to_string(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopologyDefinition {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub status: String,
    pub nodes: Vec<TopologyNode>,
    pub edges: Vec<TopologyEdge>,
    pub join_definition: Option<JoinDefinition>,
    pub cep_definition: Option<CepDefinition>,
    pub backpressure_policy: BackpressurePolicy,
    pub source_stream_ids: Vec<Uuid>,
    pub sink_bindings: Vec<ConnectorBinding>,
    pub state_backend: String,
    /// Periodic checkpoint cadence in milliseconds. The checkpoint
    /// supervisor uses this as the `tokio::time::interval` period.
    pub checkpoint_interval_ms: i32,
    /// Which runtime executes this topology. `builtin` runs in the
    /// in-process engine; `flink` is materialised as a `FlinkDeployment`
    /// CRD that the operator manages.
    pub runtime_kind: String,
    /// Job name to address when `runtime_kind = "flink"`. Required for
    /// reset/restore operations targeting Flink.
    pub flink_job_name: Option<String>,
    /// Name of the `FlinkDeployment` Custom Resource materialised by
    /// `runtime/flink/deployer.rs`. May differ from `flink_job_name`
    /// when several jobs share a session cluster.
    pub flink_deployment_name: Option<String>,
    /// Runtime job id reported by the Flink JobManager once the job is
    /// `RUNNING`. Populated by the metrics poller (D2).
    pub flink_job_id: Option<String>,
    /// Kubernetes namespace where the FlinkDeployment lives. Defaults to
    /// the value of `POD_NAMESPACE` at deployment time.
    pub flink_namespace: Option<String>,
    /// End-to-end consistency contract for this topology. Mapped to
    /// `execution.checkpointing.mode` for Flink and to the in-process
    /// engine's commit-on-checkpoint behaviour for builtin.
    pub consistency_guarantee: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateTopologyRequest {
    pub name: String,
    pub description: Option<String>,
    pub status: Option<String>,
    pub nodes: Vec<TopologyNode>,
    pub edges: Vec<TopologyEdge>,
    pub join_definition: Option<JoinDefinition>,
    pub cep_definition: Option<CepDefinition>,
    pub backpressure_policy: Option<BackpressurePolicy>,
    pub source_stream_ids: Vec<Uuid>,
    pub sink_bindings: Vec<ConnectorBinding>,
    pub state_backend: Option<String>,
    pub checkpoint_interval_ms: Option<i32>,
    pub runtime_kind: Option<String>,
    pub flink_job_name: Option<String>,
    pub flink_deployment_name: Option<String>,
    pub flink_job_id: Option<String>,
    pub flink_namespace: Option<String>,
    pub consistency_guarantee: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateTopologyRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub status: Option<String>,
    pub nodes: Option<Vec<TopologyNode>>,
    pub edges: Option<Vec<TopologyEdge>>,
    pub join_definition: Option<JoinDefinition>,
    pub cep_definition: Option<CepDefinition>,
    pub backpressure_policy: Option<BackpressurePolicy>,
    pub source_stream_ids: Option<Vec<Uuid>>,
    pub sink_bindings: Option<Vec<ConnectorBinding>>,
    pub state_backend: Option<String>,
    pub checkpoint_interval_ms: Option<i32>,
    pub runtime_kind: Option<String>,
    pub flink_job_name: Option<String>,
    pub flink_deployment_name: Option<String>,
    pub flink_job_id: Option<String>,
    pub flink_namespace: Option<String>,
    pub consistency_guarantee: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopologyRunMetrics {
    pub input_events: i32,
    pub output_events: i32,
    pub avg_latency_ms: i32,
    pub p95_latency_ms: i32,
    pub throughput_per_second: f32,
    pub dropped_events: i32,
    pub backpressure_ratio: f32,
    pub join_output_rows: i32,
    pub cep_match_count: i32,
    pub state_entries: i32,
}

impl Default for TopologyRunMetrics {
    fn default() -> Self {
        Self {
            input_events: 0,
            output_events: 0,
            avg_latency_ms: 0,
            p95_latency_ms: 0,
            throughput_per_second: 0.0,
            dropped_events: 0,
            backpressure_ratio: 0.0,
            join_output_rows: 0,
            cep_match_count: 0,
            state_entries: 0,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopologyRun {
    pub id: Uuid,
    pub topology_id: Uuid,
    pub status: String,
    pub metrics: TopologyRunMetrics,
    pub aggregate_windows: Vec<WindowAggregate>,
    pub live_tail: Vec<LiveTailEvent>,
    pub cep_matches: Vec<CepMatch>,
    pub state_snapshot: StateStoreSnapshot,
    pub backpressure_snapshot: BackpressureSnapshot,
    pub started_at: DateTime<Utc>,
    pub completed_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopologyRuntimePreview {
    pub metrics: TopologyRunMetrics,
    pub aggregate_windows: Vec<WindowAggregate>,
    pub backpressure_snapshot: BackpressureSnapshot,
    pub state_snapshot: StateStoreSnapshot,
    pub backlog_events: i32,
    pub generated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize)]
pub struct TopologyRuntimeSnapshot {
    pub topology: TopologyDefinition,
    pub latest_run: Option<TopologyRun>,
    pub preview: Option<TopologyRuntimePreview>,
    pub connector_statuses: Vec<ConnectorCatalogEntry>,
    pub latest_events: Vec<LiveTailEvent>,
    pub latest_matches: Vec<CepMatch>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ReplayTopologyRequest {
    pub stream_ids: Option<Vec<Uuid>>,
    pub from_sequence_no: Option<i64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ReplayTopologyResponse {
    pub topology_id: Uuid,
    pub stream_ids: Vec<Uuid>,
    pub replay_from_sequence_no: Option<i64>,
    pub restored_event_count: i64,
}

#[derive(Debug, Clone, FromRow)]
pub struct TopologyRow {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub status: String,
    pub nodes: SqlJson<Vec<TopologyNode>>,
    pub edges: SqlJson<Vec<TopologyEdge>>,
    pub join_definition: Option<SqlJson<JoinDefinition>>,
    pub cep_definition: Option<SqlJson<CepDefinition>>,
    pub backpressure_policy: SqlJson<BackpressurePolicy>,
    pub source_stream_ids: SqlJson<Vec<Uuid>>,
    pub sink_bindings: SqlJson<Vec<ConnectorBinding>>,
    pub state_backend: String,
    pub checkpoint_interval_ms: i32,
    pub runtime_kind: String,
    pub flink_job_name: Option<String>,
    pub flink_deployment_name: Option<String>,
    pub flink_job_id: Option<String>,
    pub flink_namespace: Option<String>,
    pub consistency_guarantee: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow)]
pub struct TopologyRunRow {
    pub id: Uuid,
    pub topology_id: Uuid,
    pub status: String,
    pub metrics: SqlJson<TopologyRunMetrics>,
    pub aggregate_windows: SqlJson<Vec<WindowAggregate>>,
    pub live_tail: SqlJson<Vec<LiveTailEvent>>,
    pub cep_matches: SqlJson<Vec<CepMatch>>,
    pub state_snapshot: SqlJson<StateStoreSnapshot>,
    pub backpressure_snapshot: SqlJson<BackpressureSnapshot>,
    pub started_at: DateTime<Utc>,
    pub completed_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<TopologyRow> for TopologyDefinition {
    fn from(value: TopologyRow) -> Self {
        Self {
            id: value.id,
            name: value.name,
            description: value.description,
            status: value.status,
            nodes: value.nodes.0,
            edges: value.edges.0,
            join_definition: value.join_definition.map(|item| item.0),
            cep_definition: value.cep_definition.map(|item| item.0),
            backpressure_policy: value.backpressure_policy.0,
            source_stream_ids: value.source_stream_ids.0,
            sink_bindings: value.sink_bindings.0,
            state_backend: value.state_backend,
            checkpoint_interval_ms: value.checkpoint_interval_ms,
            runtime_kind: value.runtime_kind,
            flink_job_name: value.flink_job_name,
            flink_deployment_name: value.flink_deployment_name,
            flink_job_id: value.flink_job_id,
            flink_namespace: value.flink_namespace,
            consistency_guarantee: value.consistency_guarantee,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

impl From<TopologyRunRow> for TopologyRun {
    fn from(value: TopologyRunRow) -> Self {
        Self {
            id: value.id,
            topology_id: value.topology_id,
            status: value.status,
            metrics: value.metrics.0,
            aggregate_windows: value.aggregate_windows.0,
            live_tail: value.live_tail.0,
            cep_matches: value.cep_matches.0,
            state_snapshot: value.state_snapshot.0,
            backpressure_snapshot: value.backpressure_snapshot.0,
            started_at: value.started_at,
            completed_at: value.completed_at,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

File: services/ingestion-replication-service/internal/domain/topology.go
package domain

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// TopologyNode mirrors event_streaming::models::topology::TopologyNode.
type TopologyNode struct {
	ID       string          `json:"id"`
	Label    string          `json:"label"`
	NodeType string          `json:"node_type"`
	StreamID *uuid.UUID      `json:"stream_id,omitempty"`
	WindowID *uuid.UUID      `json:"window_id,omitempty"`
	Config   json.RawMessage `json:"config,omitempty"`
}

// TopologyEdge mirrors event_streaming::models::topology::TopologyEdge.
type TopologyEdge struct {
	SourceNodeID string `json:"source_node_id"`
	TargetNodeID string `json:"target_node_id"`
	Label        string `json:"label"`
}

// JoinDefinition mirrors event_streaming::models::topology::JoinDefinition.
type JoinDefinition struct {
	JoinType      string    `json:"join_type"`
	LeftStreamID  uuid.UUID `json:"left_stream_id"`
	RightStreamID uuid.UUID `json:"right_stream_id"`
	TableName     string    `json:"table_name"`
	KeyFields     []string  `json:"key_fields"`
	WindowSeconds int32     `json:"window_seconds"`
}

// CepDefinition mirrors event_streaming::models::topology::CepDefinition.
type CepDefinition struct {
	PatternName   string   `json:"pattern_name"`
	Sequence      []string `json:"sequence"`
	WithinSeconds int32    `json:"within_seconds"`
	OutputStream  string   `json:"output_stream"`
}

// BackpressurePolicy mirrors event_streaming::models::topology::BackpressurePolicy.
type BackpressurePolicy struct {
	MaxInFlight      int32  `json:"max_in_flight"`
	QueueCapacity    int32  `json:"queue_capacity"`
	ThrottleStrategy string `json:"throttle_strategy"`
}

// DefaultBackpressurePolicy mirrors the Default impl on the Rust struct.
func DefaultBackpressurePolicy() BackpressurePolicy {
	return BackpressurePolicy{
		MaxInFlight:      512,
		QueueCapacity:    2048,
		ThrottleStrategy: "credit-based",
	}
}

// TopologyDefinition mirrors event_streaming::models::topology::TopologyDefinition.
type TopologyDefinition struct {
	ID                   uuid.UUID          `json:"id"`
	Name                 string             `json:"name"`
	Description          string             `json:"description"`
	Status               string             `json:"status"`
	Nodes                []TopologyNode     `json:"nodes"`
	Edges                []TopologyEdge     `json:"edges"`
	JoinDefinition       *JoinDefinition    `json:"join_definition,omitempty"`
	CepDefinition        *CepDefinition     `json:"cep_definition,omitempty"`
	BackpressurePolicy   BackpressurePolicy `json:"backpressure_policy"`
	SourceStreamIDs      []uuid.UUID        `json:"source_stream_ids"`
	SinkBindings         []ConnectorBinding `json:"sink_bindings"`
	StateBackend         string             `json:"state_backend"`
	CheckpointIntervalMS int32              `json:"checkpoint_interval_ms"`
	RuntimeKind          string             `json:"runtime_kind"`
	FlinkJobName         *string            `json:"flink_job_name,omitempty"`
	FlinkDeploymentName  *string            `json:"flink_deployment_name,omitempty"`
	FlinkJobID           *string            `json:"flink_job_id,omitempty"`
	FlinkNamespace       *string            `json:"flink_namespace,omitempty"`
	ConsistencyGuarantee string             `json:"consistency_guarantee"`
	CreatedAt            time.Time          `json:"created_at"`
	UpdatedAt            time.Time          `json:"updated_at"`
}

// TopologyRunMetrics mirrors event_streaming::models::topology::TopologyRunMetrics.
type TopologyRunMetrics struct {
	InputEvents         int32   `json:"input_events"`
	OutputEvents        int32   `json:"output_events"`
	AvgLatencyMS        int32   `json:"avg_latency_ms"`
	P95LatencyMS        int32   `json:"p95_latency_ms"`
	ThroughputPerSecond float32 `json:"throughput_per_second"`
	DroppedEvents       int32   `json:"dropped_events"`
	BackpressureRatio   float32 `json:"backpressure_ratio"`
	JoinOutputRows      int32   `json:"join_output_rows"`
	CepMatchCount       int32   `json:"cep_match_count"`
	StateEntries        int32   `json:"state_entries"`
}

// TopologyRun mirrors event_streaming::models::topology::TopologyRun.
type TopologyRun struct {
	ID                   uuid.UUID            `json:"id"`
	TopologyID           uuid.UUID            `json:"topology_id"`
	Status               string               `json:"status"`
	Metrics              TopologyRunMetrics   `json:"metrics"`
	AggregateWindows     []WindowAggregate    `json:"aggregate_windows"`
	LiveTail             []LiveTailEvent      `json:"live_tail"`
	CepMatches           []CepMatch           `json:"cep_matches"`
	StateSnapshot        StateStoreSnapshot   `json:"state_snapshot"`
	BackpressureSnapshot BackpressureSnapshot `json:"backpressure_snapshot"`
	StartedAt            time.Time            `json:"started_at"`
	CompletedAt          *time.Time           `json:"completed_at,omitempty"`
	CreatedAt            time.Time            `json:"created_at"`
	UpdatedAt            time.Time            `json:"updated_at"`
}

// TopologyRuntimePreview mirrors event_streaming::models::topology::TopologyRuntimePreview.
type TopologyRuntimePreview struct {
	Metrics              TopologyRunMetrics   `json:"metrics"`
	AggregateWindows     []WindowAggregate    `json:"aggregate_windows"`
	BackpressureSnapshot BackpressureSnapshot `json:"backpressure_snapshot"`
	StateSnapshot        StateStoreSnapshot   `json:"state_snapshot"`
	BacklogEvents        int32                `json:"backlog_events"`
	GeneratedAt          time.Time            `json:"generated_at"`
}

// TopologyRuntimeSnapshot mirrors event_streaming::models::topology::TopologyRuntimeSnapshot.
type TopologyRuntimeSnapshot struct {
	Topology          TopologyDefinition      `json:"topology"`
	LatestRun         *TopologyRun            `json:"latest_run,omitempty"`
	Preview           *TopologyRuntimePreview `json:"preview,omitempty"`
	ConnectorStatuses []ConnectorCatalogEntry `json:"connector_statuses"`
	LatestEvents      []LiveTailEvent         `json:"latest_events"`
	LatestMatches     []CepMatch              `json:"latest_matches"`
}

// ReplayTopologyRequest mirrors event_streaming::models::topology::ReplayTopologyRequest.
type ReplayTopologyRequest struct {
	StreamIDs        []uuid.UUID `json:"stream_ids,omitempty"`
	FromSequenceNo   *int64      `json:"from_sequence_no,omitempty"`
}

// ReplayTopologyResponse mirrors event_streaming::models::topology::ReplayTopologyResponse.
type ReplayTopologyResponse struct {
	TopologyID           uuid.UUID   `json:"topology_id"`
	StreamIDs            []uuid.UUID `json:"stream_ids"`
	ReplayFromSequenceNo *int64      `json:"replay_from_sequence_no,omitempty"`
	RestoredEventCount   int64       `json:"restored_event_count"`
}
